doc/man/gnunet-datastore.1
.\" This file is part of GNUnet.
.\" Copyright (C) 2001-2019 GNUnet e.V.
.\"
.\" Permission is granted to copy, distribute and/or modify this document
.\" under the terms of the GNU Free Documentation License, Version 1.3 or
.\" any later version published by the Free Software Foundation; with no
.\" Invariant Sections, no Front-Cover Texts, and no Back-Cover Texts.  A
.\" copy of the license is included in the file
.\" FDL-1.3.
.\"
.\" A copy of the license is also available from the Free Software
.\" Foundation Web site at http://www.gnu.org/licenses/fdl.html.
.\"
.\" Alternately, this document is also available under the General
.\" Public License, version 3 or later, as published by the Free Software
.\" Foundation.  A copy of the license is included in the file
.\" GPL3.
.\"
.\" A copy of the license is also available from the Free Software
.\" Foundation Web site at http://www.gnu.org/licenses/gpl.html.
.\"
.\" SPDX-License-Identifier: GPL3.0-or-later OR FDL1.3-or-later
.\"
.Dd April 13, 2017
.Dt GNUNET-DATASTORE 1
.Os
.Sh NAME
.Nm gnunet-datastore
.Nd dump or insert (restore) GNUnet datastore databases
.Sh SYNOPSIS
.Nm
.Op Fl c Ar FILENAME | Fl -config= Ns Ar FILENAME
.Op Fl d | -dump
.Op Fl f Ar FILENAME | Fl -file= Ns Ar FILENAME
.Op Fl h | -help
.Op Fl i | -insert
.Op Fl L Ar LOGLEVEL | Fl -loglevel= Ns Ar LOGLEVEL
.Op Fl l Ar FILENAME | Fl -logfile= Ns Ar FILENAME
.Op Fl v | -version
.Sh DESCRIPTION
.Nm
can be used to backup and restore or merge GNUnet datastores.
This is useful if a datastore is to be migrated between
SQL databases, i.e. from sqlite to postgres or vice versa.
gnunet-datastore will dump the entire contents of the database
or insert a dump file into the database.
.Bl -tag -width indent
.It Fl c Ar FILENAME | Fl -config= Ns Ar FILENAME
Configuration file to use
.It Fl d | -dump
Dump all records to a file
.It Fl f Ar FILENAME | Fl -file= Ns Ar FILENAME
File to dump to or insert from.
Otherwise stdin/stdout are used.
.It Fl h | -help
Print the help page.
.It Fl i | -insert
Insert from dump file.
.It Fl L Ar LOGLEVEL | Fl -loglevel= Ns Ar LOGLEVEL
Change the loglevel.
Possible values for LOGLEVEL are ERROR, WARNING, INFO and DEBUG.
.It Fl l Ar FILENAME | Fl -logfile= Ns Ar FILENAME
Configure logging to write logs to FILENAME
.It Fl v | -version
Print the version number
.El
.\".Sh EXAMPLES
.Sh SEE ALSO
.Xr gnunet-publish 1
.Pp
The full documentation for gnunet is maintained as a Texinfo manual.
If the
.Xr info 1
and gnunet programs are properly installed at your site, the command
.Pp
.Dl info gnunet
.Pp
should give you access to the complete handbook,
.Pp
.Dl info gnunet-c-tutorial
.Pp
will give you access to a tutorial for developers.
.sp
Depending on your installation, this information is also available in
.Xr gnunet 7 and
.Xr gnunet-c-tutorial 7 .
.\".Sh HISTORY
.\".Sh AUTHORS
.Sh BUGS
Report bugs by using
.Lk https://bugs.gnunet.org
or by sending electronic mail to
.Aq Mt gnunet-developers@gnu.org .


ChangeLog
Wed Oct 30 10:50:00 UTC 2019
  Released gnunet 0.11.8
  -ng0

Sun Oct 27 17:17:00 2019 +0200
  Released gnunet 0.11.7
  -schanzen, ng0

Fri Oct 25 00:00:00 UTC 2019
  Dropped glpk dependency(-check). -ng0

Wed Oct 23 20:55:12 2019 +0200
  Make REST service start per user in a multiuser setup.
  -schanzen

Wed Oct 23 00:00:00 UTC 2019
  Remove setuid helpers. They never worked as intended. The
  fixed version is not portable and the defacto good way to
  handle this which is most portable is to let package managers
  handle this (as they do right now, the good ones), and let
  people handle this who read the documentation.
  This commit removes what would be patched out by the majority
  of package managers, which is setuid handling in Makefiles.
  It is very likely that no one will notice this code is gone.
  -ng0

Thu Oct 17 00:00:00 UTC 2019
  Added support for doas, use it in some places conditionally
  if sudo is not present. Fixed sudo detection. -ng0

Mon Oct 14 00:00:00 UTC 2019
  Added 'pretty' make rule to run uncrustify over the source tree.
  -ng0

Mon Sep 16 00:00:00 UTC 2019
  Remove checks for CXX and CPP. Scans found no
  (remaining) cxx code. ccache with clang on
  NetBSD 9 hit no rebuilds. -ng0

Mon Sep 10 00:00:00 UTC 2019
  Drop win32 and mingw support after discussion with
  Christian. It has been unmaintained in our code base
  for years and there are no known users.
  Future implementors for Windows 10 support can
  refer to this commit as reference. -ng0

Mon Sep 9 00:00:00 UTC 2019
  Remove plibc, closes #5877 as discussed in
  https://lists.gnu.org/archive/html/gnunet-developers/2019-09/msg00002.html
  -ng0

Sun Sep 1 13:25:07 2019 +0200
  Improved support for external applications/services wrt
  localization/gettext.
  https://lists.gnu.org/archive/html/gnunet-developers/2019-09/msg00000.html

Tue 20 Aug 2019 04:32:09 PM CEST
  Fixed fatal bug in gnunet-bcd that prevented the main page
  from showing up. -CG

Mon Aug 19 17:36:48 2019 +0200
  REST: Added API plugin to read/write GNUnet configuration options -alexia

Sun Jul 28 00:00:00 2019 +0000
  Adding --enable-texinfo4 configure switch, defaulting to off.
  This allows distributors to override the texinfo 5.x and later
  commandline options we use, and allows us on the serverside to
  build html with newer output options. -ng0

Wed 24 Jul 2019 05:43:43 PM CEST
  Releasing GNUnet 0.11.6. -CG

Mon Jul 15 09:57:20 2019 +0200
  The REST server now has a configurable CORS Origin setting -schanzen

Fri Jul 5 22:11:04 2019 +0200
  Added detection of SUID binaries
  through GNUNET_OS_get_suid_binary_path -lurchi

Sat 29 Jun 2019 10:18:55 AM CEST
  Renamed "heap" namestore backend to "flat" as it does
  persist data in a file. Made it work with databases larger
  than 16 MB. -CG

Fri Jun 28 14:10:53 2019 +0200
  RECLAIM: Reissue of tickets (same attributes, same identity)
  will no longer create a new ticket, but reuse the old. -alexia

Thu Jun 27 10:34:46 2019 +0200
  The REST server now returns the Origin header value in the
  CORS header reponse. -schanzen

Thu Jun 27 09:08:06 2019 +0200
  GNS NSS plugin will not timeout and not hang forever and block
  gethostbyname. Final resolution pending #5782 -schanzen

Sat Jun 22 10:38:43 2019 +0200
  Add gnutls support for GNS CA certificate generation -schanzen

Sat 15 Jun 2019 11:45:35 PM CEST
  Improved gnunet-gns performance for many ego scenario. -CG

Fri 14 Jun 2019 07:17:40 PM CEST
  Add option to gnunet-identity to display private keys. -CG

Wed Jun 5 21:19:46 2019 +0200
 Releasing GNUnet 0.11.5. -schanzen

Wed May 29 11:17:49 2019 +0200
  Added CAA record type (RFC 6844). -schanzen

Sat 25 May 2019 08:27:39 AM CEST
  Significantly speed-up ego creation and deletion via
  gnunet-identity command when there are many (hundreds)
  of egos at the same peer. -CG

Sun May 12 12:00:00 2019 +0200
  Releasing GNUnet 0.11.4. -MS

Thu May 9 23:35:24 2019 +0200
  gnunet-identity can now output the public key on creation of new
  egos, and -e can act as a filter when listing all egos. -CG

Thu May 2 08:44:37 2019 +0200
  gnunet-arm now logs to file by default -CG

Sat Apr 27 21:50:57 2019 +0200
  RECLAIM: Added ECDH encryption for authorization code payload -schanzen

Sat 27 Apr 2019 06:36:31 AM CEST
  Bump libmicrohttpd version requirement as some intermediate
  versions have relevant bugs that cause problems. -CG

Sat Apr 27 01:50:43 2019 +0200
  RECLAIM: No longer experimental -schanzen

Fri Apr 26 11:08:21 2019 +0200
  REST: Configuration plugin added -schanzen

Tue Apr 23 23:00:00 UTC 2019
  All manpages are properly licensed. -ng0

Sun Apr 21 22:22:22 UTC 2019
  All manpages are now in mdoc format. -ng0

Sat Apr 20 18:38:43 UTC 2019
  Remove optional gnunet-download-manager.scm and
  with it the optional dependency on a no longer
  maintained version of GNU Guile (#5682). -ng0

Sun Apr 14 09:54:20 2019 +0200
  RECLAIM: Moved away from ABE to GNS encryption. Minor API changed and
  renaming of records. Deprecated SQLite plugin in favor of namestore.
  -schanzen

Sun Apr 7 14:40:59 2019 +0200
  Releasing GNUnet 0.11.3. -CG

Sat 06 Apr 2019 02:36:17 PM CEST
  Introducing GNUNET_memcmp() and GNUNET_is_zero() for better typing
  when comparing memory areas. -CG

Fri 05 Apr 2019 11:38:53 PM CEST
  Use paths of sysctl, ip, ifconfig and ip(6)tables from configure
  if they work. (#5352). -CG

Thu 04 Apr 2019 12:39:44 PM CEST
  Hotfix for build issue. Releasing GNUnet 0.11.2. -CG/MS

Wed 03 Apr 2019 03:18:43 PM CEST
  Releasing GNUnet 0.11.1. -CG

Wed 03 Apr 2019 01:54:44 PM CEST
  Merged gnunet-qr-c-2 branch, replacing Python-based implementation
  of gnunet-qr with C-based implementation that also integrates
  gnunet-uri. -HG/MS/CG

Wed 03 Apr 2019 01:15:57 PM CEST
  Fixed a few test case regressions (GNS, FS). -CG

Mon Apr 1 22:00:00 UTC 2019
  documentation/buildsystem: manpages and other documentation parts
  can now be build more controlled again, this fixes the regression

src/datastore/gnunet-datastore.c
/*
     This file is part of GNUnet.
     Copyright (C) 2010, 2013 GNUnet e.V.

     GNUnet is free software: you can redistribute it and/or modify it
     under the terms of the GNU Affero General Public License as published
     by the Free Software Foundation, either version 3 of the License,
     or (at your option) any later version.

     GNUnet is distributed in the hope that it will be useful, but
     WITHOUT ANY WARRANTY; without even the implied warranty of
     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
     Affero General Public License for more details.

     You should have received a copy of the GNU Affero General Public License
     along with this program.  If not, see <http://www.gnu.org/licenses/>.

     SPDX-License-Identifier: AGPL3.0-or-later
 */

/**
 * @file datastore/gnunet-datastore.c
 * @brief tool to manipulate datastores
 * @author Christian Grothoff
 */
#include <inttypes.h>
#include "platform.h"
#include "gnunet_util_lib.h"
#include "gnunet_datastore_service.h"

GNUNET_NETWORK_STRUCT_BEGIN

struct DataRecord
{
  /**
   * Number of bytes in the item (NBO).
   */
  uint32_t size GNUNET_PACKED;

  /**
   * Type of the item (NBO) (actually an enum GNUNET_BLOCK_Type)
   */
  uint32_t type GNUNET_PACKED;

  /**
   * Priority of the item (NBO).
   */
  uint32_t priority GNUNET_PACKED;

  /**
   * Desired anonymity level (NBO).
   */
  uint32_t anonymity GNUNET_PACKED;

  /**
   * Desired replication level (NBO).
   */
  uint32_t replication GNUNET_PACKED;

  /**
   * Expiration time (NBO).
   */
  struct GNUNET_TIME_AbsoluteNBO expiration;

  /**
   * Key under which the item can be found.
   */
  struct GNUNET_HashCode key;
};
GNUNET_NETWORK_STRUCT_END


/**
 * Length of our magic header.
 */
static const size_t MAGIC_LEN = 16;

/**
 * Magic header bytes.
 */
static const uint8_t MAGIC_BYTES[16] = "GNUNETDATASTORE1";

/**
 * Dump the database.
 */
static int dump;

/**
 * Insert into the database.
 */
static int insert;

/**
 * Dump file name.
 */
static char *file_name;

/**
 * Dump file handle.
 */
static struct GNUNET_DISK_FileHandle *file_handle;

/**
 * Global return value.
 */
static int ret;

/**
 * Handle for datastore.
 */
static struct GNUNET_DATASTORE_Handle *datastore;

/**
 * Current operation.
 */
static struct GNUNET_DATASTORE_QueueEntry *qe;

/**
 * Record count.
 */
static uint64_t record_count;


static void
do_shutdown (void *cls)
{
  if (NULL != qe)
    GNUNET_DATASTORE_cancel (qe);
  if (NULL != datastore)
    GNUNET_DATASTORE_disconnect (datastore, GNUNET_NO);
  if (NULL != file_handle)
    GNUNET_DISK_file_close (file_handle);
}


/**
 * Begin dumping the database.
 */
static void
start_dump (void);


/**
 * Begin inserting into the database.
 */
static void
start_insert (void);


/**
 * Perform next GET operation.
 */
static void
do_get (const uint64_t next_uid);


/**
 * Process a datum that was stored in the datastore.
 *
 * @param cls closure
 * @param key key for the content
 * @param size number of bytes in data
 * @param data content stored
 * @param type type of the content
 * @param priority priority of the content
 * @param anonymity anonymity-level for the content
 * @param replication replication-level for the content
 * @param expiration expiration time for the content
 * @param uid unique identifier for the datum;
 *        maybe 0 if no unique identifier is available
 */
static void
get_cb (void *cls,
        const struct GNUNET_HashCode *key,
        size_t size,
        const void *data,
        enum GNUNET_BLOCK_Type type,
        uint32_t priority,
        uint32_t anonymity,
        uint32_t replication,
        struct GNUNET_TIME_Absolute expiration,
        uint64_t uid)
{
  qe = NULL;
  if (NULL == key)
  {
    fprintf (stderr, _ ("Dumped %" PRIu64 " records\n"), record_count);
    GNUNET_DISK_file_close (file_handle);
    file_handle = NULL;
    if (insert)
      start_insert ();
    else
    {
      ret = 0;
      GNUNET_SCHEDULER_shutdown ();
    }
    return;
  }

  struct DataRecord dr;
  dr.size = htonl ((uint32_t) size);
  dr.type = htonl (type);
  dr.priority = htonl (priority);
  dr.anonymity = htonl (anonymity);
  dr.replication = htonl (replication);
  dr.expiration = GNUNET_TIME_absolute_hton (expiration);
  dr.key = *key;

  ssize_t len;
  len = GNUNET_DISK_file_write (file_handle, &dr, sizeof(dr));
  if (sizeof(dr) != len)
  {
    fprintf (stderr,
             _ ("Short write to file: %zd bytes expecting %zd\n"),
             len,
             sizeof(dr));
    ret = 1;
    GNUNET_SCHEDULER_shutdown ();
    return;
  }

  len = GNUNET_DISK_file_write (file_handle, data, size);
  if (size != len)
  {
    fprintf (stderr,
             _ ("Short write to file: %zd bytes expecting %zd\n"),
             len,
             size);
    ret = 1;
    GNUNET_SCHEDULER_shutdown ();
    return;
  }

  record_count++;
  do_get (uid + 1);
}


/**
 * Perform next GET operation.
 */
static void
do_get (const uint64_t next_uid)
{
  GNUNET_assert (NULL == qe);
  qe = GNUNET_DATASTORE_get_key (datastore,
                                 next_uid,
                                 false /* random */,
                                 NULL /* key */,
                                 GNUNET_BLOCK_TYPE_ANY,
                                 0 /* queue_priority */,
                                 1 /* max_queue_size */,
                                 &get_cb,
                                 NULL /* proc_cls */);
  if (NULL == qe)
  {
    fprintf (stderr, _ ("Error queueing datastore GET operation\n"));
    ret = 1;
    GNUNET_SCHEDULER_shutdown ();
  }
}


/**
 * Begin dumping the database.
 */
static void
start_dump ()
{
  record_count = 0;

  if (NULL != file_name)
  {
    file_handle = GNUNET_DISK_file_open (file_name,
                                         GNUNET_DISK_OPEN_WRITE
                                         | GNUNET_DISK_OPEN_TRUNCATE
                                         | GNUNET_DISK_OPEN_CREATE,
                                         GNUNET_DISK_PERM_USER_READ
                                         | GNUNET_DISK_PERM_USER_WRITE);
    if (NULL == file_handle)
    {
      fprintf (stderr, _ ("Unable to open dump file: %s\n"), file_name);
      ret = 1;
      GNUNET_SCHEDULER_shutdown ();
      return;
    }
  }
  else
  {
    file_handle = GNUNET_DISK_get_handle_from_int_fd (STDOUT_FILENO);
  }
  GNUNET_DISK_file_write (file_handle, MAGIC_BYTES, MAGIC_LEN);
  do_get (0);
}


/**
 * Continuation called to notify client about result of the
 * operation.
 *
 * @param cls closure
 * @param success GNUNET_SYSERR on failure (including timeout/queue drop)
 *                GNUNET_NO if content was already there
 *                GNUNET_YES (or other positive value) on success
 * @param min_expiration minimum expiration time required for 0-priority content to be stored
 *                by the datacache at this time, zero for unknown, forever if we have no
 *                space for 0-priority content
 * @param msg NULL on success, otherwise an error message
 */
static void
put_cb (void *cls,
        int32_t success,
        struct GNUNET_TIME_Absolute min_expiration,
        const char *msg)
{
  qe = NULL;
  if (GNUNET_SYSERR == success)
  {
    fprintf (stderr, _ ("Failed to store item: %s, aborting\n"), msg);
    ret = 1;
    GNUNET_SCHEDULER_shutdown ();
    return;
  }

  struct DataRecord dr;
  ssize_t len;

  len = GNUNET_DISK_file_read (file_handle, &dr, sizeof(dr));
  if (0 == len)
  {
    fprintf (stderr, _ ("Inserted %" PRIu64 " records\n"), record_count);
    ret = 0;
    GNUNET_SCHEDULER_shutdown ();
    return;
  }
  else if (sizeof(dr) != len)
  {
    fprintf (stderr,
             _ ("Short read from file: %zd bytes expecting %zd\n"),
             len,
             sizeof(dr));
    ret = 1;
    GNUNET_SCHEDULER_shutdown ();
    return;
  }

  const size_t size = ntohl (dr.size);
  uint8_t data[size];
  len = GNUNET_DISK_file_read (file_handle, data, size);
  if (size != len)
  {
    fprintf (stderr,
             _ ("Short read from file: %zd bytes expecting %zd\n"),
             len,
             size);
    ret = 1;
    GNUNET_SCHEDULER_shutdown ();
    return;
  }

  record_count++;
  qe = GNUNET_DATASTORE_put (datastore,
                             0,
                             &dr.key,
                             size,
                             data,
                             ntohl (dr.type),
                             ntohl (dr.priority),
                             ntohl (dr.anonymity),
                             ntohl (dr.replication),
                             GNUNET_TIME_absolute_ntoh (dr.expiration),
                             0,
                             1,
                             &put_cb,
                             NULL);
  if (NULL == qe)
  {
    fprintf (stderr, _ ("Error queueing datastore PUT operation\n"));
    ret = 1;
    GNUNET_SCHEDULER_shutdown ();
  }
}


/**
 * Begin inserting into the database.
 */
static void
start_insert ()
{
  record_count = 0;

  if (NULL != file_name)
  {
    file_handle = GNUNET_DISK_file_open (file_name,
                                         GNUNET_DISK_OPEN_READ,
                                         GNUNET_DISK_PERM_NONE);
    if (NULL == file_handle)
    {
      fprintf (stderr, _ ("Unable to open dump file: %s\n"), file_name);
      ret = 1;
      GNUNET_SCHEDULER_shutdown ();
      return;
    }
  }
  else
  {
    file_handle = GNUNET_DISK_get_handle_from_int_fd (STDIN_FILENO);
  }

  uint8_t buf[MAGIC_LEN];
  ssize_t len;

  len = GNUNET_DISK_file_read (file_handle, buf, MAGIC_LEN);
  if ((len != MAGIC_LEN) || (0 != memcmp (buf, MAGIC_BYTES, MAGIC_LEN)))
  {
    fprintf (stderr, _ ("Input file is not of a supported format\n"));
    return;
  }
  put_cb (NULL, GNUNET_YES, GNUNET_TIME_UNIT_ZERO_ABS, NULL);
}


/**
 * Main function that will be run by the scheduler.
 *
 * @param cls closure
 * @param args remaining command-line arguments
 * @param cfgfile name of the configuration file used
 * @param cfg configuration
 */
static void
run (void *cls,
     char *const *args,
     const char *cfgfile,
     const struct GNUNET_CONFIGURATION_Handle *cfg)
{
  GNUNET_SCHEDULER_add_shutdown (&do_shutdown, NULL);
  datastore = GNUNET_DATASTORE_connect (cfg);
  if (NULL == datastore)
  {
    fprintf (stderr, _ ("Failed connecting to the datastore.\n"));
    ret = 1;
    GNUNET_SCHEDULER_shutdown ();
    return;
  }
  if (dump)
    start_dump ();
  else if (insert)
    start_insert ();
  else
  {
    fprintf (stderr,
             _ ("Please choose at least one operation: %s, %s\n"),
             "dump",
             "insert");
    ret = 1;
    GNUNET_SCHEDULER_shutdown ();
  }
}


/**
 * The main function to manipulate datastores.
 *
 * @param argc number of arguments from the command line
 * @param argv command line arguments
 * @return 0 ok, 1 on error
 */
int
main (int argc, char *const *argv)
{
  struct GNUNET_GETOPT_CommandLineOption options[] =
  { GNUNET_GETOPT_option_flag ('d',
                               "dump",
                               gettext_noop (
                                 "Dump all records from the datastore"),
                               &dump),
    GNUNET_GETOPT_option_flag ('i',
                               "insert",
                               gettext_noop (
                                 "Insert records into the datastore"),
                               &insert),
    GNUNET_GETOPT_option_filename ('f',
                                   "file",
                                   "FILENAME",
                                   gettext_noop ("File to dump or insert"),
                                   &file_name),
    GNUNET_GETOPT_OPTION_END };

  if (GNUNET_OK != GNUNET_STRINGS_get_utf8_args (argc, argv, &argc, &argv))
    return 2;

  if (GNUNET_OK !=
      GNUNET_PROGRAM_run (argc,
                          argv,
                          "gnunet-datastore",
                          gettext_noop ("Manipulate GNUnet datastore"),
                          options,
                          &run,
                          NULL))
    ret = 1;
  GNUNET_free ((void *) argv);
  return ret;
}


/* end of gnunet-datastore.c */
